src/layer.rs

```diff
@@ -1,102 +1,732 @@
-use serde::Serialize;
+use std::fmt;
 
-#[derive(Debug, Clone, Default, Serialize)]
-#[serde(rename_all = "camelCase")]
-pub struct Layer {
+use serde::{Deserialize, Serialize};
+use wasm_bindgen::{closure::Closure, JsValue};
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum Expression<T> {
+    Enum(T),
+    String(String),
+    Number(f64),
+    Bool(bool),
+    List(Vec<Expression<T>>),
+}
+
+pub trait EnumMarker {}
+impl<T: EnumMarker> From<T> for Expression<T> {
+    fn from(item: T) -> Self {
+        Expression::Enum(item)
+    }
+}
+impl<T> From<&str> for Expression<T> {
+    fn from(value: &str) -> Self {
+        Expression::String(value.into())
+    }
+}
+
+impl<T> From<String> for Expression<T> {
+    fn from(value: String) -> Self {
+        Expression::String(value)
+    }
+}
+
+impl<T> From<f64> for Expression<T> {
+    fn from(value: f64) -> Self {
+        Expression::Number(value)
+    }
+}
+
+impl<T> From<bool> for Expression<T> {
+    fn from(value: bool) -> Self {
+        Expression::Bool(value)
+    }
+}
+
+impl<T> From<Vec<&str>> for Expression<T> {
+    fn from(value: Vec<&str>) -> Self {
+        Expression::List(value.into_iter().map(|s| s.into()).collect())
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+#[serde(rename_all = "kebab-case")]
+pub enum Visibility {
+    #[default]
+    Visible,
+    None,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct GetLayer {
     pub id: String,
-    #[serde(skip_serializing_if = "Option::is_none")]
-    pub layout: Option<Layout>,
+    #[serde(rename = "type")]
+    pub type_: String,
     #[serde(skip_serializing_if = "Option::is_none")]
     pub maxzoom: Option<f64>,
     #[serde(skip_serializing_if = "Option::is_none")]
     pub minzoom: Option<f64>,
     #[serde(skip_serializing_if = "Option::is_none")]
-    pub paint: Option<Paint>,
+    pub visibility: Option<Visibility>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase", tag = "type")]
+pub enum Layer {
+    Custom(CustomLayer),
+    Fill(FillLayer),
+    Background(BackgroundLayer),
+    Line(LineLayer),
+    Symbol(SymbolLayer),
+    Circle(CircleLayer),
+}
+
+pub trait IntoLayer {
+    fn into_layer(self) -> Layer;
+}
```

Comment:
Now we get to the heart of the matter. A strongly typed implementation of the layers. Every layer implement the `IntoLayer` trait so that these layers can all be passed into `map.add_layer()`. By using the serde annotations we can actually deserialize the output of `getStyle` and get out all correct strongly types layers.

It also allows for individual `Paint` and `Layout` properties. That is really exciting because it means you get autocomplete on all of those features! Which is super nice in my opinion. I'm just confused about why they haven't done it like that in the Javascript/Typesciprt implementation and I'm wondering if I'm really missing something that invalidates this approach? Let me know what you think.